libs/contrib/Data/Fin/Extra.idr

```diff
@@ -97,3 +123,49 @@ divMod {ok=_} (S n) (S d) =
         Right (r' ** eq') => Fraction {ok=ok} (S n) (S d) q (FS r') $
             rewrite sym $ plusSuccRightSucc (q * S d) (finToNat r') in
                 cong S $ trans (sym $ cong (plus (q * S d)) eq') eq
+
+-------------------------------------------------
+--- Splitting operations and their properties ---
+-------------------------------------------------
+
+export
+splitSumFin : {a : Nat} -> Fin (a + b) -> Either (Fin a) (Fin b)
+splitSumFin {a=Z}   x      = Right x
+splitSumFin {a=S k} FZ     = Left FZ
+splitSumFin {a=S k} (FS x) = bimap FS id $ splitSumFin x
+
+export
+splitProdFin : {a, b : Nat} -> Fin (a * b) -> (Fin a, Fin b)
+splitProdFin {a=S _} x = case splitSumFin x of
+  Left  y => (FZ, y)
+  Right y => bimap FS id $ splitProdFin y
+
+--- Properties ---
+
+export
+splitSumFin_correctness : {a, b : Nat} -> (x : Fin $ a + b) ->
+                          case splitSumFin {a} {b} x of
+                            Left  l => x = weakenN b l
+                            Right r => x = shift a r
+splitSumFin_correctness {a=Z}   x  = Refl
+splitSumFin_correctness {a=S k} FZ = Refl
+splitSumFin_correctness {a=S k} (FS x) with (splitSumFin_correctness x)
+  splitSumFin_correctness {a=S k} (FS x) | subcorr with (splitSumFin x)
+    splitSumFin_correctness {a=S k} (FS x) | subcorr | Left  y = rewrite subcorr in Refl
+    splitSumFin_correctness {a=S k} (FS x) | subcorr | Right y = rewrite subcorr in Refl
+
+export
+splitProdFin_correctness : {a, b : Nat} -> (x : Fin $ a * b) ->
+                           let (o, i) = splitProdFin {a} {b} x in
+                           finToNat x = finToNat o * b + finToNat i
```

Comment:
Can we not define the other direction `Fin a -> Fin b -> Fin (a * b)`
rather than going via `finToNat` for the correctness lemma?

Reply:
I'm sure yes, let me poke a bit

Reply:
Well, after implementing an appropriate (as I thought) function `Fin a -> Fin b -> Fin (a * b)` (which I dared to add to this PR), I realized that `splitProdFin` is actually a `divMod` function for division of `Fin (a * b)` by `(b : Nat)`.

So, there exists a function that does reverse, i.e. multiplying `Fin a` by `(b : Nat)` adding `Fin b` as a remainder, resulting to `Fin (a * b)`, but its correctness shall be proven also through `finToNat` because it acts on the border of `Fin`s and `Nat`s.

Reply:
Well, I understand now, that we actually can try to reformulate correctness property not through `finToNat`, but using some `Nat` to `Fin` operations, like `last`. That is, instead of `finToNat x = finToNat o * b + finToNat i` we could write something like `x = i + o * last` having appropriate operations `+` and `*` upon `Fin`s (like those that I added recently). (I'm not sure there're no off-by-one errors in the example)

Do you think that this kind of proof has more value than `finToNat`-based?

Reply:
Well, that was tough, but possible: https://github.com/idris-lang/Idris2/pull/830/commits/6ddbb6da3d1d1f6a23a4a7378b3086f4482b96c4. Is this approximately what you were originally talking about?